server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectWriteAheadLogs.java

```diff
@@ -83,13 +85,17 @@ public class GarbageCollectWriteAheadLogs {
     this.fs = fs;
     this.liveServers = liveServers;
     this.walMarker = new WalStateManager(context);
-    this.store = () -> Iterators.concat(
-        context.getAmple().readTablets().forLevel(DataLevel.ROOT).filter(new HasWalsFilter())
-            .fetch(LOCATION, LAST, LOGS, PREV_ROW, SUSPEND).build().iterator(),
-        context.getAmple().readTablets().forLevel(DataLevel.METADATA).filter(new HasWalsFilter())
-            .fetch(LOCATION, LAST, LOGS, PREV_ROW, SUSPEND).build().iterator(),
-        context.getAmple().readTablets().forLevel(DataLevel.USER).filter(new HasWalsFilter())
-            .fetch(LOCATION, LAST, LOGS, PREV_ROW, SUSPEND).build().iterator());
+    TabletsMetadata root = context.getAmple().readTablets().forLevel(DataLevel.ROOT)
+        .filter(new HasWalsFilter()).fetch(LOCATION, LAST, LOGS, PREV_ROW, SUSPEND).build();
+    TabletsMetadata metadata = context.getAmple().readTablets().forLevel(DataLevel.METADATA)
+        .filter(new HasWalsFilter()).fetch(LOCATION, LAST, LOGS, PREV_ROW, SUSPEND).build();
+    TabletsMetadata user = context.getAmple().readTablets().forLevel(DataLevel.USER)
+        .filter(new HasWalsFilter()).fetch(LOCATION, LAST, LOGS, PREV_ROW, SUSPEND).build();
+    this.store = Streams.concat(root.stream(), metadata.stream(), user.stream()).onClose(() -> {
```

Comment:
I think the way this is structured the GC will only be able to run one cycle and then will crash or not see data (not sure which) when attempting to run subsequent cycles.  The code uses this store to create an iterator over the data each time the gc runs a cycle.  So it should not be creating the streams in the constructor, but instead it should create them each time on an as needed basis.  So maybe store could be changed to a `Supplier<Stream<TabletMetadata>>`.

I think the existing design of this code around this `store` is kinda tricky. I was fighting with this same code in #4196 and made the same mistake of creating stuff in the constructor until I caught it in testing.  AFAICT this design is driven by the desire for test to inject the `store`, was not really sure how to improve it when working on #4196 because of the test.  Would like to figure how to make this code less misleading.

Reply:
From what I can tell each time a `GarbageCollectWriteAheadLogs` object is created, it is used to call `collect` just once (which eventually iterates over the `store`) but it does not seem like it is ever reused. Is your concern with if collect was called again on the same `GarbageCollectWriteAheadLogs` object? Is that something we want to allow for? If we don't see the need to ever allow that to happen we could add some sort of check to make sure its only called once in which case it seems like it might be fine to keep things how they are. I may be thinking about this wrong though

Reply:
Another option could be to keep things the same (using the iterators as the underlying resource for `store`) then make `GarbageCollectWriteAheadLogs` autoclosable and make sure the resources are tracked and closed when the object is. Then we could do something like this:

```diff
-            GarbageCollectWriteAheadLogs walogCollector =
-                new GarbageCollectWriteAheadLogs(getContext(), fs, liveTServerSet);
-            log.info("Beginning garbage collection of write-ahead logs");
-            walogCollector.collect(status);
+            try(GarbageCollectWriteAheadLogs walogCollector =
+                new GarbageCollectWriteAheadLogs(getContext(), fs, liveTServerSet)) {
+              log.info("Beginning garbage collection of write-ahead logs");
+              walogCollector.collect(status);
+            }
```

Reply:
Could also add the check that was mentioned earlier to gaurd against collect being called more than once.

Reply:
I took the suggested approach in 8212985. 

I then thought of the idea to make getters for the things we want to inject from the test, that way we dont have to create the store in the constructor and can more cleanly close the resources without having to make the object autoclosable. This approach was taken in bd4e81a. Let me know which approach looks better.